feat: Adds support for JavaScript AppSync Functions

GiamPy5 opened the pull request:
## Description
Currently the module does not support the newly released JavaScript runtime for AppSync Functions. Recently, the JS runtime was introduced (in this module) for resolvers, but not for functions.

Furthermore, I've added support for `sync_config` configuration.

## Motivation and Context
Resolves [Enable JS functions](#48)

## Breaking Changes
None.

## How Has This Been Tested?
- [X] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [X] I have tested and validated these changes using one or more of the provided `examples/*` projects
<!--- Users should start with an existing example as its written, deploy it, then check their changes against it -->
<!--- This will highlight breaking/disruptive changes. Once you have checked, deploy your changes to verify -->
<!--- Please describe how you tested your changes -->
- [X] I have executed `pre-commit run -a` on my pull request
<!--- Please see https://github.com/antonbabenko/pre-commit-terraform#how-to-install for how to install -->

Commit:
feat: Adds support for JavaScript AppSync Functions

GiamPy5 commented:
Any feedback on this?

GiamPy5 commented:
@antonbabenko for visibility :)

juanoi commented:
looking forward to this being released!

antonbabenko commented on main.tf:
```diff
@@ -3,6 +3,11 @@ locals {
     field = join(".", slice(split(".", k), 1, length(split(".", k))))
   }) if var.create_graphql_api }
+
+  functions = { for k, v in var.functions : k => merge(v, {
```
Why do we need this for `functions`? I think it will work well without it.

GiamPy5 replied:
Makes sense. I've simplified it.

antonbabenko commented on main.tf:
```diff
@@ -221,16 +226,44 @@ resource "aws_appsync_resolver" "this" {
+  request_mapping_template  = lookup(each.value, "request_mapping_template", try(each.value.runtime.name == "APPSYNC_JS", false) ? null : "{}")
+  response_mapping_template = lookup(each.value, "response_mapping_template", try(each.value.runtime.name == "APPSYNC_JS", false) ? null : "{}")
+
+  code = try(each.value.runtime.name == "APPSYNC_JS", false) ? each.value.code : null
```
```suggestion
  code = try(each.value.code, null)
```

antonbabenko commented on main.tf:
```diff
@@ -221,16 +226,44 @@ resource "aws_appsync_resolver" "this" {
+      conflict_handler   = try(sync_config.value.conflict_handler, "NONE")
+
+      dynamic "lambda_conflict_handler_config" {
+        for_each = try(sync_config.value.conflict_handler == "LAMBDA", false) ? try([each.value.sync_config.lambda_conflict_handler_config], []) : []
```
```suggestion
        for_each = try([each.value.sync_config.lambda_conflict_handler_config], [])
```

Let users figure out what they need to specify. Also, we have this shown in the example.

antonbabenko commented on examples/complete/schema.graphql:
```diff
@@ -12,10 +12,16 @@ type User {
   name: String
 }
 
+type Entity {
```
Let's remove `Entity` from schema and use existing types to illustrate the functionality of JS functions on type `User`, for example.

antonbabenko commented:
Please verify examples by running them yourself. I just did and got this error:

```
╷
│ Error: creating AppSync Function: BadRequestException: SyncConfig can only be specified for a resolver attached to a DynamoDB data source.
│ {
│   RespMetadata: {
│     StatusCode: 400,
│     RequestID: "1686ceb4-8d3f-46d0-8fc2-62120eba051f"
│   },
│   Message_: "SyncConfig can only be specified for a resolver attached to a DynamoDB data source."
│ }
│ 
│   with module.appsync.aws_appsync_function.this["JavaScript_Complete"],
│   on ../../main.tf line 223, in resource "aws_appsync_function" "this":
│  223: resource "aws_appsync_function" "this" {

```

GiamPy5 commented:
PR has been updated, I've fixed the examples and I've added a `account_recovery_setting` configuration in the cognito pool to avoid terraform throwing the following error when modifying the pool:

```
│ Error: updating Cognito User pool (eu-west-1_sUsTbzu9Y): InvalidParameter: 1 validation error(s) found.
│ - minimum field size of 1, UpdateUserPoolInput.AccountRecoverySetting.RecoveryMechanisms.
│ 
│ 
│   with aws_cognito_user_pool.this,
│   on main.tf line 300, in resource "aws_cognito_user_pool" "this":
│  300: resource "aws_cognito_user_pool" "this" {
│ 
```

antonbabenko commented:
Thanks @GiampaoloFalqui !

antonbabenko commented:
This PR is included in version 1.8.0 :tada:
